File: app/models/audition.rb
class Audition < ApplicationRecord
  require 'csv'
  include PgSearch::Model
  pg_search_scope :search_by,
                  against: [:firstname, :lastname, :email, :status, :genre, :created_at, :id, :artist_name],
                  using: {
                    tsearch: {prefix: true}
                  }

  GENRES = ['Acapella', 'Acid', 'Jazz', 'Acoustic', 'Acid Punk', 'Alternative', 'Hip Hop', 'Ambient',
            'Avantgarde', 'Bass', 'Blues', 'Cabaret', 'Celtic Chamber', 'Chanson', 'Chorus', 'Christian Rap',
            'Cinematic', 'Classical', 'Classic Rock', 'Club', 'Comedy', 'Country', 'Cult'].freeze
  SOURCES = ['Facebook', 'Instagram', 'Twitter', 'Other'].freeze
  PENDING = "Pending".freeze
  ACCEPTED = "Accepted".freeze
  REJECTED = "Rejected".freeze
  DELETED = "Deleted".freeze
  STATUSES = [PENDING, ACCEPTED, REJECTED, DELETED].freeze
  CSV_ATTRIBUTES = %w[id name artist_name email genres formatted_created_at assigned_to status].freeze
  NAME_REGEX = /(?=.*[[:^alnum:]])/.freeze
  AUDITION_ACCEPT_EMAIL="Hi [name]!Thanks for submitting music to Audiosocket, we have listened to your link and would love to work with you! Please sign in to our artist portal here. There, you can update your artist profile, submit music, artworks, etc. The more assets you can give us the better. Once submitted, our team will review and will start the ingestion into our catalog.Thanks! Music Licensing Coordinator".freeze
  AUDITION_REJECT_EMAIL="Hello [name],Thank you for submitting an audition. After careful review, we have decided the music you submitted is not a match for our current needs. Please understand, that while your music may not be a match this time, it certainly might be in the future as our clients’ needs are constantly changing.With that in mind, we encourage you to submit new music again in the future.Have a great day!".freeze

  enum status: STATUSES

  validates :firstname, length: { maximum: 30 }, format: { without: NAME_REGEX }
  validates :lastname, length: { maximum: 30 }, format: { without: NAME_REGEX }

  has_many :links, dependent: :destroy

  accepts_nested_attributes_for :links

  after_initialize :default_email_status, if: :new_record?

  scope :by_status, ->(status) { where(status: status) }
  scope :count_by_status, ->(status) { by_status(status).count }

  def self.to_csv
    CSV.generate(headers: true) do |csv|
      csv << CSV_ATTRIBUTES
      all.each do |audition|
        csv <<  audition.csv_row
      end
    end
  end

  def csv_row
    CSV_ATTRIBUTES.map{ |attribute| send(attribute) }
  end

  def self.search(query, sort, direction, status)
    scope = self.all
    scope = scope.search_by(query) if query.present?
    scope = scope.order("#{sort} #{direction}") if sort.present?
    scope = scope.by_status(status) if status.present?

    scope
  end

  def default_email_status
    self.status ||= PENDING
  end

  def name
    "#{firstname} #{lastname}"
  end

  def genres
    genre.join(', ')
  end

  def formatted_created_at
    created_at.strftime("%d %B %y %I:%M %p")
  end

  def manager_assigned(audition)
    audition.assigned_to
  end
end

File: app/javascript/packs/audition.js
$(document).ready(function() {
  $("#other-text-field").addClass("display-none");
  $('#media').on('change', function() {
    let conceptName = $('#media').find(":selected").text();
    if (conceptName == "Other") {
      $("#other-text-field").removeClass("display-none");
    } else {
      $("#other-text-field").addClass("display-none");
    }
  });

  function check_to_hide_or_show_add_link() {
    if ($('#link-fields .nested-fields:visible').length == 4) {
      $('#link-fields .links').hide();
    } else {
      $('#link-fields .links').show();
   }
  }

  $('#link-fields').on('cocoon:after-insert', function() {
   check_to_hide_or_show_add_link();
  });

  $('#link-fields').on('cocoon:after-remove', function() {
   check_to_hide_or_show_add_link();
  });

  check_to_hide_or_show_add_link();

  $('.select').on('change', function() {
    $.ajax({
      url: '/assigned_to_update',
      type:'GET',
      url: '/update',
      type:'PATCH',
      data: {
        audition_id: $(this).attr("id"),
        id: $(this).attr("id"),
        assigned_to: this.value,
      }
    });
  });
});

$(document).on("click", ".modal-footer a", function(){
  let aud_id = $(this).attr("id");
  let val = $(this).attr("value");
  let email_content ;
  if(val == "Accepted"){
    let accept_email = "Hi [name],Thanks for submitting music to Audiosocket. We have listened to your link and would love to work with you. Please sign in to our artist portal here. There, you can update your artist profile, submit music, artworks, etc. The more assets you can give us the better. Once submitted, our team will review and will start the ingestion into our catalog.Thanks! Music Licensing Coordinator."
    let desc = CKEDITOR.instances['email_description'].setData(accept_email);
    $(document).on("click", ".btn-email", function(){
      desc = CKEDITOR.instances['email_description'].getData();
      email_content = desc.slice(3,-5);
      ajaxRequest(aud_id,val,email_content);
    });
  }
  else if(val == "Rejected"){
    let reject_email = "Hello [name], Thank you for submitting an audition. After careful review, we have decided the music you submitted is not a match for our current needs. Please understand, that while your music may not be a match this time, it certainly might be in the future as our clients’ needs are constantly changing. With that in mind, we encourage you to submit new music again in the future. Have a great day!"
    let desc = CKEDITOR.instances['email_description'].setData(reject_email);
    $(document).on("click", ".btn-email", function(){
      desc = CKEDITOR.instances['email_description'].getData();
      email_content = desc.slice(3,-5);
      ajaxRequest(aud_id,val,email_content);
    });
  }
  else if(val == "Deleted"){
    ajaxRequest(aud_id,val);
  }

  function ajaxRequest(aud_id, val, email_content){
    let audition_id = aud_id;
    let status = val;
    let email_description = email_content ;
    $.ajax({
      url: '/update_status_email',
      type:'PATCH',
      data: {
        id: audition_id,
        status: status,
        email_description: email_description
      },
      success: function(data){
        location.reload();
      }
    });
  }
});

